### PR TITLE
The ETF Trick: Generate a series of returns given a set of instruments

### PR DESCRIPTION
**Pull request by joshvera**
Added an implementation of The ETF Trick to model a set of instruments given a dynamic set of weightings. I've updated `Portfolio.ipynb` with example usage.

The resulting time series reflects the value of $1 invested in a set of instruments. The series is strictly positive and can take rebalance costs, bid-ask spread costs, and missing data into account, though some work remains to extend it to those use cases.

To extend this beyond cash-settled instruments, one can also model a basket of futures as if they were a single non-expiring cash product with this method, but I think we'd need another function to handle the rolls of those futures contracts.

This idea comes from _Advances in Financial Machine Learning_ by Marco Lopez de Prado.

### CODE REVIEW

**Comment by joshvera** [comment]
I 🍩 know what that test failure is about.

File: model.py [hunk 1]
```diff
@@ -452,36 +453,46 @@ def fetchQuote(self, instrument: Instrument) -> Quote:
         pass
 
 
+@dataclass
```
Comment (jspahrsummers) [inline model.py]: Oh, thank fuck. Thanks for pointing this out. Want to do the same for the other models? :pray:

Comment (jspahrsummers) [inline model.py]:
```suggestion
@dataclass(frozen=True)
```

File: model.py [hunk 2]
```diff
@@ -452,36 +453,46 @@ def fetchQuote(self, instrument: Instrument) -> Quote:
...
     @classmethod
     def quantizeQuantity(cls, quantity: Decimal) -> Decimal:
         return quantity.quantize(cls.quantityQuantization,
                                  rounding=ROUND_HALF_EVEN)
 
-    def __init__(self, instrument: Instrument, quantity: Decimal,
-                 costBasis: Cash):
-        if instrument.currency != costBasis.currency:
+    @classmethod
+    def averagePrice(cls, instrument, quantity, costBasis) -> Cash:
```
Comment (jspahrsummers) [inline model.py]: Testing locally, `mypy` appears to be barfing because this method is named the same as the `averagePrice` attribute.

Why is it now an attribute, anyway? We don't want it to be a required argument for initialization.

File: model.py [hunk 3]
```diff
@@ -452,36 +453,46 @@ def fetchQuote(self, instrument: Instrument) -> Quote:
...
 
-        if not quantity.is_finite():
+        if not self.quantity.is_finite():
             raise ValueError(
-                'Position quantity {} is not a finite number'.format(quantity))
+                'Position quantity {} is not a finite number'.format(
+                    self.quantity))
 
-        quantity = self.quantizeQuantity(quantity)
+        quantity = self.quantizeQuantity(self.quantity)
```
Comment (jspahrsummers) [inline model.py]:
```suggestion
        self.quantity = self.quantizeQuantity(self.quantity)
```

… and omit the assignment below

File: model.py [hunk 4]
```diff
@@ -493,35 +504,6 @@ def combine(self, other: 'Position') -> 'Position':
...
-            assert self.costBasis == 0
-            return self.costBasis
-
-        return self.costBasis / self.quantity / self.instrument.multiplier
-
-    @property
-    def costBasis(self) -> Cash:
-        return self._costBasis
-
     def __repr__(self) -> str:
```
Comment (jspahrsummers) [inline model.py]: Can ditch this now, as `@dataclass` generates it.

File: ibkr.py [hunk 5]
```diff
@@ -122,11 +122,14 @@ def extractPosition(p: IB.Position) -> Position:
...
         costBasis = parseFiniteDecimal(p.avgCost) * qty
+        basis = Cash(currency=Currency[p.contract.currency],
+                     quantity=costBasis)
 
         return Position(instrument=instrument,
                         quantity=qty,
-                        costBasis=Cash(currency=Currency[p.contract.currency],
-                                       quantity=costBasis))
+                        costBasis=basis,
+                        averagePrice=Position.averagePrice(
```
Comment (jspahrsummers) [inline ibkr.py]: As commented on `model.py`, why is this required now?

File: etf.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+from dataclasses import asdict
+from ib_insync import util
+import ib_insync as ibapi
+
+
+def etf(portfolio: pd.DataFrame) -> pd.Series:
+    """
+    Returns a time series representing the investment of $1 in a basket of instruments weighted proportionally by the given weights.
+
+    @param val: A DataFrame of instruments containing Open, Close, and Weight data indexed by Date.
```
Comment (jspahrsummers) [inline etf.py]:
```suggestion
    @param portfolio: A DataFrame of instruments containing Open, Close, and Weight data indexed by Date.
```

File: etf.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    """
+    # Initialize a zero'd out T-sized array where T is the length of the date range.
+    etf = np.zeros(portfolio.loc['open'].shape[0])
```
Comment (jspahrsummers) [inline etf.py]: Are DataFrames' columns case-sensitive? You called it `Open` above.

File: etf.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    etf[0] = Decimal(1)  # Initial AUM for this instrument is $1.
+
+    # Initialize an I x T array of holdings over time so we can keep track of how we've divided our AUM
+    # across the given instruments over time.
+    hodls = np.zeros(portfolio.loc['open'].shape)
```
Comment (jspahrsummers) [inline etf.py]:
```suggestion
    holds = np.zeros(portfolio.loc['open'].shape)
```
?

File: etf.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    return (prices / prices.shift(1) - 1)[1:]
+
+
+def positions_to_dataframe(positions: List[model.Position]) -> pd.DataFrame:
+    """
+    Returns a dataframe of positions with an additional `value` and `allocation` columns
+    calculated from the averagePrice and quantity.
+    """
+    is_stock = lambda position: type(position.instrument) == model.Stock
+    stocks = list(map(lambda p: asdict(p), filter(is_stock, positions)))
```
Comment (jspahrsummers) [inline etf.py]: TIOLI but I mostly find list comprehensions more readable:

```suggestion
    stocks = [asdict(p) for p in positions if type(position.instrument) == model.Stock]
```

File: etf.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    for i, row in frame.iterrows():
+        weights[row['instrument'].symbol] = row['allocation']
+        components[row['instrument'].symbol] = historical_data[i]
+
+    portfolio = stocks_to_portfolio(components, weights)
+
+    return portfolio_to_returns(portfolio, timezone)
+
+
+def position_dataframe_to_history(ib: ibapi.IB,
```
Comment (jspahrsummers) [inline etf.py]: Can you parameterize this with the `LiveDataProvider` type from `model`, extending it as needed? (We can rename it to just `DataProvider`, probably, given the context here.)

Want to keep this stuff relatively generic in case we add support for other data sources (or brokers with data APIs) later.

File: etf.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+    return list(map(util.df, bars))
+
+
+def holdings(val: pd.DataFrame, hodls: np.ndarray, i: pd.DataFrame, t: int,
+             aum_t: int) -> Decimal:
+    open_price = val[i].loc['open'][t]
+
+    # If the open price is NaN, this instrument's open wasn't recorded at time t.
+    # So let's use the previous day's calculation.
+    if math.isnan(open_price):
```
Comment (jspahrsummers) [inline etf.py]:
```suggestion
    if not open_price.is_finite():
```

File: etf.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        # TODO: make the exchange rate flexible. This should generally be the dollar value of 1 point of instrument i.
+        exchange_rate = Decimal(1)
+        # The purpose of the (weights_i_t * sum_of_weights ^ -1) in the holdings calculation is to de-lever the allocations.
+        sum_of_weights: Decimal = val.loc['weight'].iloc[t].abs().sum()
+
+        # If today was t + 1, one way to calculate today's holdings would be by using the opening price, since for a future
+        # we may not know the closing price of a new contract at roll time.
+        # If open prices are unavailable, then the last close price at t will work too.
+        next_open = val[i].loc['close'][t]
+        if math.isnan(next_open):
```
Comment (jspahrsummers) [inline etf.py]:
```suggestion
        if not next_open.is_finite():
```

File: etf.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+def prices_to_daily_returns(prices: pd.Series) -> pd.Series:
+    """
+    Calculates daily returns for a Series of prices.
+    Note: drops the first value of the given Series.
+    """
+    return (prices / prices.shift(1) - 1)[1:]
+
+
+def positions_to_dataframe(positions: List[model.Position]) -> pd.DataFrame:
```
Comment (jspahrsummers) [inline etf.py]:
```suggestion
def positions_to_dataframe(positions: Iterable[model.Position]) -> pd.DataFrame:
```

File: etf.py (hunk 10 above)
Comment (joshvera) [reply etf.py]: I think parameterizing by `LiveDataProvider` is outside the scope of this PR.

While extending `LiveDataProvider` is [straightforward](https://github.com/erdewit/ib_insync/blob/1fa68f3594872584556cc23afc4f5e032ce24212/ib_insync/ib.py#L878), reconciling the returned data isn't unless you'd like to write interfaces for those as well. See [ib_insync/objects.py](https://github.com/erdewit/ib_insync/blob/1fa68f3594872584556cc23afc4f5e032ce24212/ib_insync/objects.py#L116) for an example.

File: model.py (hunk 1 above)
Comment (joshvera) [reply model.py]: We can't due to mutating `quantity` internally because of `quantizeQuantity`. Opted for `unsafe_hash=True` to generate the `__hash__` function instead.

File: model.py (hunk 2 above)
Comment (joshvera) [reply model.py]: This was due to trying to construct a dictionary with it. Opted to add keys to it instead https://github.com/jspahrsummers/bankroll/pull/20/files#diff-6c78c5f8cdc75208a344d8a38cc08959R77.

File: etf.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    weights = {}
+    components = {}
+    for i, row in frame.iterrows():
+        weights[row['instrument'].symbol] = row['allocation']
+        components[row['instrument'].symbol] = historical_data[i]
+
+    return stocks_to_portfolio(components, weights)
+
+
+def positions_to_history(provider: model.LiveDataProvider,
```
Comment (jspahrsummers) [inline etf.py]: FWIW got this error on my portfolio:

> Error 162, reqId 14: Historical Market Data Service error message:No market data permissions for BATEUK, CHIXUK, LSE, TRQXUK STK, contract: Stock(conId=35151400, symbol='GAW', exchange='SMART', primaryExchange='LSE', currency='GBP', localSymbol='GAW', tradingClass='GAW')

I intentionally don't have a data subscription for LSE, but still have a security there. Does that mean historical data is unavailable? (If so, fine, but we should filter out those failed lookups here.)

File: etf.py [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+import pyfolio as pf
+from typing import Optional, List, Dict, Iterable
+from decimal import Decimal
+import model
+from dataclasses import asdict
+from ib_insync import util
+import ib_insync as ibapi
+
+
+def etf(portfolio: pd.DataFrame, timezone: str) -> pd.Series:
```
Comment (jspahrsummers) [inline etf.py]: Thinking out loud… I know these pandas types are meant to be dynamic, but would something like [`NewType`](https://docs.python.org/3.7/library/typing.html?highlight=typing#newtype) make sense, so at least we avoid mixing them incorrectly? (Not for this PR, just considering how to make it safer over time.)

File: ibkr.py [hunk 16]
```diff
@@ -420,6 +422,22 @@ def __init__(self, client: IB.IB):
         self._client = client
         super().__init__()
 
+    def qualifyContract(self, instrument: Instrument) -> IB.Contract:
```
Comment (jspahrsummers) [inline ibkr.py]: Can you update `fetchQuote()` below to use this as well?

File: .circleci/config.yml [hunk 17]
```diff
@@ -17,7 +17,7 @@ jobs:
     docker:
       # specify the version you desire here
       # use `-browsers` prefix for selenium tests, e.g. `3.6.1-browsers`
-      - image: circleci/python:3.6.1
+      - image: circleci/python:3.7.2-stretch
```
[the remark this answers is not in the document]
Comment (joshvera) [reply .circleci/config.yml]: @dataclasses is only in 3.7+.
